src/csharp/Grpc.Core/Internal/ChannelCredentialsSafeHandle.cs

```diff
@@ -15,11 +15,16 @@
 #endregion
 using System;
 using System.Runtime.InteropServices;
-using System.Threading;
-using System.Threading.Tasks;
 
 namespace Grpc.Core.Internal
 {
+    internal delegate int VerifyPeerCallbackInternal(
```

Comment:
I think the targetHost and targetPem need to be IntPtr and you need to convert them to a string yourself (again see my example). the problem is that the automatic char* to string conversion might do copying or deallocation of the char* and that's something you don't want.

https://github.com/grpc/grpc/blob/master/src/csharp/Grpc.Core/Internal/NativeMetadataCredentialsPlugin.cs#L65

Reply:
Yes you're right thanks, fixed it.